File: client/game_card/Assets/Scripts/Core/Log/Log.cs

namespace Utopia
{
    public class Log
    {
        protected ILogImpl m_logImpl = null;
        protected object m_param;
        
        public bool Init(ILogImpl logImpl, object param)
        {
            m_logImpl = logImpl;
            m_param = param;
            return m_logImpl.Init(param);
        }

        public void Release()
        {
            if (null != m_logImpl)
            {
                m_logImpl.Release();
                m_logImpl = null;
            }
        }

        LogLevel m_logLvl = LogLevel.Debug;
        public void SetLogLvl(LogLevel lvl)
        {
            m_logLvl = lvl;
        }
        public LogLevel logLvl { get { return m_logLvl; } }

        public void Debug(string format, params object[] args)
        {
            this.DoLog(LogLevel.Debug, format, args);
        }
        public void Info(string format, params object[] args)
        {
            this.DoLog(LogLevel.Info, format, args);
        }

        public void Warning(string format, params object[] args)
        {
            this.DoLog(LogLevel.Waring, format, args);
        }

        public void Error(string format, params object[] args)
        {
            this.DoLog(LogLevel.Error, format, args);
        }

        public void Exception(System.Exception e)
        {
            this.DoLog(LogLevel.Exception, "{0}", e.ToString());
        }

        public void Assert(bool isTrue, string format, params object[] args)
        {
            if (!isTrue)
            {
                this.DoLog(LogLevel.Assert, format, args);
            }
        }

        public void DoLog(LogLevel lvl, string format, params object[] args)
        {
            if (lvl < m_logLvl)
                return;

            m_logImpl.DoLog(lvl, string.Format(format, args));
        }

        public void DoLogContent(LogLevel lvl, string content)
        {
            if (lvl < m_logLvl)
                return;

            m_logImpl.DoLog(lvl, content);
        }
    }
}

File: client/game_card/LuaScript/libs/log.lua

local is_enable_traceback = false
function log_set_enable_traceback(is_enable)
    is_enable_traceback = is_enable
end

function string_format(fmt_str, ...)
    local str_list = {}
    local tb = {...}
    local tb_len = select('#', ...)
    for i=1, tb_len do
        local arg = tb[i]
        if "table" ~= type(arg) then
            table.insert(str_list, tostring(arg))
        else
            local Show_Max_Level = 5
            table.insert(str_list, string.toprint(arg, Show_Max_Level))
        end
    end
    local ret = fmt_str
    if #str_list > 0 then
        ret = string.format(fmt_str, table.unpack(str_list))
    end
    if is_enable_traceback then
        ret = debug.traceback(ret)
    end
    return ret
end

function log_set_level(log_lvl)
    CS.Utopia.AppLog.SetLogLvl(log_lvl)
end

function log_debug(fmt_str, ...)
    fmt_str = string.format("debug::%s", fmt_str)
    local log_str = string_format(fmt_str, ...)

    -- string.format("debug::%s", log_str)

    CS.Utopia.AppLog.DoLogContent(CS.Utopia.LogLevel.Debug, log_str)
end

function log_info(fmt_str, ...)
    local log_str = string_format(fmt_str, ...)
    CS.Utopia.AppLog.DoLogContent(CS.Utopia.LogLevel.Info, log_str)
end

function log_warn(fmt_str, ...)
    local log_str = string_format(fmt_str, ...)
    CS.Utopia.AppLog.DoLogContent(CS.Utopia.LogLevel.Waring, log_str)
end

function log_error(fmt_str, ...)
    local log_str = string_format(fmt_str, ...)
    log_str = debug.traceback(log_str)
    CS.Utopia.AppLog.DoLogContent(CS.Utopia.LogLevel.Error, log_str)
end

function log_assert(is_ok, fmt_str, ...)
    if not is_ok then
        log_error(fmt_str, ...)
        assert(false)
    end
end

old_print = print

function print(...)
    local str_list = {}
    for _, arg in pairs({...}) do
        if "table" ~= type(arg) then
            table.insert(str_list, tostring(arg))

        else
            table.insert(str_list, string.to_print(arg, 3))
        end
    end
    local ret = table.concat(str_list, ' ')
    -- old_print(ret)
    log_debug(ret)
end

function log_print(...)
    print(...)
end

